File: apps/Pose-Motion/iOS/PoseMotion/Pipeline/PosePipeline.swift
import CoreGraphics
import CoreMedia
import CoreVideo
import Foundation
import ZeticMLange

struct FrameResult {
    let image: CGImage?
    let frameSize: CGSize
    let personBox: CGRect?          // normalized
    let ballBox: CGRect?            // normalized
    let keypoints: [Keypoint2D]?    // COCO-17, normalized frame coords
    let pose3D: [SIMD3<Float>]?     // H36M-17, root-relative
    let timings: FrameTimings
}

/// The three-model chain: YOLO26n (person + ball) → RTMPose (crop → 2D skeleton)
/// → MotionBERT-Lite (keypoint window → 3D). Everything runs on one serial queue.
final class PosePipeline {
    let queue = DispatchQueue(label: "ai.zetic.posemotion.inference")
    let detector = MelangeModel(label: "YOLO26n")
    let pose = MelangeModel(label: "RTMPose-s")
    let lift = MelangeModel(label: "MotionBERT-Lite")

    /// Fires on the main thread with the latest processed frame.
    var onResult: ((FrameResult, BenchmarkSnapshot) -> Void)?

    private let preprocessor = FramePreprocessor()
    private let simcc = SimCCDecoder()
    private let window = KeypointWindow()
    private let stats = BenchmarkStats()
    private var frameCounter = 0
    private var lastKeypoints: [Keypoint2D]?
    private var lastPose3D: [SIMD3<Float>]?

    func resetSession() {
        queue.async { [weak self] in
            self?.window.reset()
            self?.frameCounter = 0
            self?.lastKeypoints = nil
            self?.lastPose3D = nil
        }
    }

    /// Must be called on `queue` (the ClipFrameSource pump already is).
    func process(_ pixelBuffer: CVPixelBuffer, at time: CMTime) {
        let t0 = CFAbsoluteTimeGetCurrent()
        var timings = FrameTimings()
        let frameSize = CGSize(
            width: CVPixelBufferGetWidth(pixelBuffer),
            height: CVPixelBufferGetHeight(pixelBuffer)
        )

        // 1. Detector: person + ball
        var person: Detection?
        var ball: Detection?
        if detector.isLoaded, let input = preprocessor.detectorTensor(from: pixelBuffer) {
            if let outputs = try? detector.run([input]), let out = outputs.first {
                (person, ball) = DetectorDecoder.decode(out)
                timings.detectMs = detector.lastLatencyMs
            }
        }

        // 2. Pose: crop → SimCC → keypoints
        var keypoints: [Keypoint2D]?
        if pose.isLoaded, let personBox = person?.rect {
            let crop = PersonCropper.cropRect(for: personBox, frameSize: frameSize)
            if crop.width > 4, crop.height > 4,
               let input = preprocessor.poseTensor(from: pixelBuffer, cropPixels: crop),
               let outputs = try? pose.run([input]), outputs.count >= 2 {
                // Identify simcc_x vs simcc_y by bin count, not output order.
                let a = outputs[0], b = outputs[1]
                let aBins = a.shape.last ?? 0
                let bBins = b.shape.last ?? 0
                let (sx, sy) = aBins <= bBins ? (a, b) : (b, a)   // 384 x-bins < 512 y-bins
                keypoints = simcc.decode(simccX: sx, simccY: sy,
                                         cropPixels: crop, frameSize: frameSize)
                timings.poseMs = pose.lastLatencyMs
            }
        }

        // 3. Keypoint window (repeat last frame on a miss so the lift window stays coherent)
        if let kps = keypoints {
            window.push(kps, frameSize: frameSize)
            lastKeypoints = kps
        } else if let last = lastKeypoints {
            window.push(last.map { Keypoint2D(x: $0.x, y: $0.y, conf: 0.1) }, frameSize: frameSize)
        }

        // 4. 3D lift, every liftStride frames
        if lift.isLoaded, frameCounter % AppConfig.liftStride == 0,
           window.hasData, let input = window.tensorData() {
            if let outputs = try? lift.run([input]), let out = outputs.first {
                lastPose3D = Self.readPose3D(out)
                timings.liftMs = lift.lastLatencyMs
            }
        }
        frameCounter += 1

        timings.totalMs = (CFAbsoluteTimeGetCurrent() - t0) * 1000
        stats.record(timings)

        let result = FrameResult(
            image: preprocessor.displayImage(from: pixelBuffer),
            frameSize: frameSize,
            personBox: person?.rect,
            ballBox: ball?.rect,
            keypoints: keypoints,
            pose3D: lastPose3D,
            timings: timings
        )
        let snapshot = stats.snapshot()
        DispatchQueue.main.async { [weak self] in
            self?.onResult?(result, snapshot)
        }
    }

    /// [1,T,17,3] → the configured window frame as 17 root-relative 3D joints.
    private static func readPose3D(_ output: Tensor) -> [SIMD3<Float>]? {
        let shape = output.shape
        guard shape.count == 4, shape[2] == 17, shape[3] == 3 else { return nil }
        let t = min(AppConfig.liftReadIndex, shape[1] - 1)
        let floats: [Float] = output.data.withUnsafeBytes { Array($0.bindMemory(to: Float.self)) }
        let base = t * 17 * 3
        guard floats.count >= base + 17 * 3 else { return nil }
        return (0..<17).map { j in
            SIMD3(floats[base + j * 3], floats[base + j * 3 + 1], floats[base + j * 3 + 2])
        }
    }
}

File: apps/Pose-Motion/Android/app/src/main/java/ai/zetic/demo/posemotion/pipeline/PosePipeline.kt
package ai.zetic.demo.posemotion.pipeline

import ai.zetic.demo.posemotion.AppConfig
import ai.zetic.demo.posemotion.benchmark.BenchmarkStats
import ai.zetic.demo.posemotion.video.DecodedFrame
import android.graphics.RectF
import com.zeticai.mlange.core.tensor.DataType
import com.zeticai.mlange.core.tensor.Tensor
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * The three-model chain: YOLO26n (person + ball) → RTMPose (crop → 2D skeleton)
 * → MotionBERT-Lite (keypoint window → 3D). Everything runs on the inference thread.
 * Line-for-line port of the iOS PosePipeline.
 */
class PosePipeline {
    val detector = MelangeModel("YOLO26n")
    val pose = MelangeModel("RTMPose-s")
    val lift = MelangeModel("MotionBERT-Lite")

    /** Called on the inference thread with each processed frame. */
    var onResult: ((FrameResult, BenchmarkSnapshot) -> Unit)? = null

    private val preprocessor = FramePreprocessor()
    private val window = KeypointWindow()
    private val stats = BenchmarkStats()
    private var frameCounter = 0
    private var lastKeypoints: List<Keypoint2D>? = null
    private var lastPose3D: List<Vec3>? = null

    private val liftBuffer: ByteBuffer = ByteBuffer
        .allocateDirect(AppConfig.LIFT_WINDOW * 17 * 3 * 4)
        .order(ByteOrder.nativeOrder())

    fun resetSession() {
        window.reset()
        frameCounter = 0
        lastKeypoints = null
        lastPose3D = null
        stats.resetFps()
    }

    /** One-time first inference per model on zero inputs, so the rolling stats
     *  never contain the backend compile spike. Inference thread only. */
    fun warmup() {
        if (detector.isLoaded) {
            val zeros = ByteBuffer.allocateDirect(3 * 640 * 640 * 4).order(ByteOrder.nativeOrder())
            runCatching { detector.run(arrayOf(Tensor(zeros, DataType.Float32, intArrayOf(1, 3, 640, 640)))) }
        }
        if (pose.isLoaded) {
            val zeros = ByteBuffer.allocateDirect(3 * 256 * 192 * 4).order(ByteOrder.nativeOrder())
            runCatching { pose.run(arrayOf(Tensor(zeros, DataType.Float32, intArrayOf(1, 3, 256, 192)))) }
        }
        if (lift.isLoaded) {
            val zeros = ByteBuffer.allocateDirect(81 * 17 * 3 * 4).order(ByteOrder.nativeOrder())
            runCatching { lift.run(arrayOf(Tensor(zeros, DataType.Float32, intArrayOf(1, 81, 17, 3)))) }
        }
    }

    fun process(frame: DecodedFrame) {
        val t0 = System.nanoTime()
        val timings = FrameTimings()
        val bitmap = frame.bitmap
        val frameW = bitmap.width.toFloat()
        val frameH = bitmap.height.toFloat()

        // 1. Detector: person + ball
        var person: Detection? = null
        var ball: Detection? = null
        if (detector.isLoaded) {
            runCatching {
                val outputs = detector.run(arrayOf(preprocessor.detectorTensor(bitmap)))
                val data = MelangeModel.floats(outputs[0])
                val decoded = DetectorDecoder.decode(data)
                person = decoded.first
                ball = decoded.second
                timings.detectMs = detector.lastLatencyMs
            }
        }

        // 2. Pose: crop → SimCC → keypoints
        var keypoints: List<Keypoint2D>? = null
        val personBox = person?.rect
        if (pose.isLoaded && personBox != null) {
            val crop = PersonCropper.cropRect(personBox, frameW, frameH)
            if (crop.width() > 4 && crop.height() > 4) {
                runCatching {
                    val outputs = pose.run(arrayOf(preprocessor.poseTensor(bitmap, crop)))
                    if (outputs.size >= 2) {
                        // Identify simcc_x vs simcc_y by size (Tensor shape is private):
                        // x has 17*384 = 6528 floats, y has 17*512 = 8704.
                        val a = MelangeModel.floats(outputs[0])
                        val b = MelangeModel.floats(outputs[1])
                        val (sx, sy) = if (a.size <= b.size) a to b else b to a
                        keypoints = SimCCDecoder.decode(sx, sy, crop, frameW, frameH)
                        timings.poseMs = pose.lastLatencyMs
                    }
                }
            }
        }

        // 3. Keypoint window (repeat last frame on a miss so the lift window stays coherent)
        val kps = keypoints
        if (kps != null) {
            window.push(kps, frameW, frameH)
            lastKeypoints = kps
        } else {
            lastKeypoints?.let { last ->
                window.push(last.map { Keypoint2D(it.x, it.y, 0.1f) }, frameW, frameH)
            }
        }

        // 4. 3D lift, every LIFT_STRIDE frames
        if (lift.isLoaded && frameCounter % AppConfig.LIFT_STRIDE == 0 && window.hasData) {
            window.windowFloats()?.let { floats ->
                runCatching {
                    liftBuffer.clear()
                    liftBuffer.asFloatBuffer().put(floats)
                    val input = Tensor(
                        liftBuffer, DataType.Float32,
                        intArrayOf(1, AppConfig.LIFT_WINDOW, 17, 3)
                    )
                    val outputs = lift.run(arrayOf(input))
                    readPose3D(MelangeModel.floats(outputs[0]))?.let { lastPose3D = it }
                    timings.liftMs = lift.lastLatencyMs
                }
            }
        }
        frameCounter++

        timings.totalMs = (System.nanoTime() - t0) / 1_000_000.0
        stats.record(timings)
        if (frameCounter % 30 == 0) {
            val s = stats.snapshot()
            android.util.Log.d(
                "stats",
                "[stats] fps=%.1f det=%.1f pose=%.1f lift=%.1f total=%.1f mem=%.0f"
                    .format(s.fps, s.detectMs, s.poseMs, s.liftMs, s.totalMs, s.memoryMB)
            )
        }

        val result = FrameResult(
            bitmap = bitmap,
            frameWidth = bitmap.width,
            frameHeight = bitmap.height,
            personBox = personBox,
            ballBox = ball?.rect,
            keypoints = keypoints,
            pose3D = lastPose3D,
            timings = timings,
        )
        onResult?.invoke(result, stats.snapshot())
    }

    fun deinitAll() {
        detector.deinit()
        pose.deinit()
        lift.deinit()
    }

    /** [1,T,17,3] flat floats → the configured window frame as 17 root-relative joints. */
    private fun readPose3D(floats: FloatArray): List<Vec3>? {
        if (floats.size % 51 != 0) return null
        val frames = floats.size / 51
        if (frames == 0) return null
        val t = minOf(AppConfig.LIFT_READ_INDEX, frames - 1)
        val base = t * 51
        return (0 until 17).map { j ->
            Vec3(floats[base + j * 3], floats[base + j * 3 + 1], floats[base + j * 3 + 2])
        }
    }
}
